File: model/rules.py
"""
탐지 규칙 (규칙 문서 [2] 개별 규칙, [3] 조합 규칙)

각 규칙 = (조건 + 임계값) → RuleHit | None
임계값/가중치는 config.py에서 주입.

"""
from __future__ import annotations

import re
from datetime import datetime
from typing import Optional

from config import RISK_CONFIG as CFG
from models import ComboHit, RuleHit, Transaction, UserBaseline

# ---------- each rules ----------

def rule_r1(tx: Transaction, base: UserBaseline) -> Optional[RuleHit]:
    """R1 고액 이체: r = amount / avg_transfer"""
    if tx.type != "transfer" or base.avg_transfer <= 0:
        return None
    r = tx.amount / base.avg_transfer
    c = CFG["r1"]
    if r >= c["t3"]:
        weight, reason = c["w3"], "평소 대비 극단적 고액 이체 (사기 전형)"
    elif r >= c["t2"]:
        weight, reason = c["w2"], "평소 이체액의 7배 이상"
    elif r >= c["t1"]:
        weight, reason = c["w1"], "평소보다 큰 금액 이체"
    else:
        return None
    return RuleHit("R1", "고액 이체", weight, reason, {"r": r})


def rule_r2(tx: Transaction, base: UserBaseline) -> Optional[RuleHit]:
    """R2 신규 수취인: 이체 수취계좌가 known_payees에 없음"""
    if tx.type != "transfer" or not tx.payee_account:
        return None
    if tx.payee_account in base.known_payees:
        return None
    return RuleHit("R2", "신규 수취인", CFG["r2"]["weight"], "처음 거래하는 계좌")


def rule_r3(tx: Transaction, base: UserBaseline) -> Optional[RuleHit]:
    """R3 비활동시간 거래: 거래 시각이 active_hours 밖"""
    hour = _local_hour(tx.timestamp)
    start, end = base.active_hours
    active = (start <= hour <= end) if start <= end else (hour >= start or hour <= end)
    if active:
        return None
    return RuleHit("R3", "비활동시간 거래", CFG["r3"]["weight"], "평소 거래하지 않는 시간대")


def rule_r4(tx: Transaction, recent: list[Transaction]) -> Optional[RuleHit]:
    """R4 단기 연속 거래: 이체/출금 반복. 최댓값 1건만 적용"""
    def is_send(t: Transaction) -> bool:
        return t.type in ("transfer", "withdrawal")

    if not is_send(tx):
        return None

    now = _epoch(tx.timestamp)
    prior = [_epoch(t.timestamp) for t in recent if is_send(t)]

    def count_within(minutes: int) -> int:
        # recent는 현재 거래 미포함 → +1
        return sum(1 for t in prior if 0 <= now - t <= minutes * 60) + 1

    c = CFG["r4"]
    if count_within(c["long"]["window_min"]) >= c["long"]["count"]:
        return RuleHit("R4", "단기 연속 거래", c["long"]["weight"], "30분 내 5건 이상 반복 송·출금")
    if count_within(c["short"]["window_min"]) >= c["short"]["count"]:
        return RuleHit("R4", "단기 연속 거래", c["short"]["weight"], "10분 내 3건 이상 반복 송·출금")
    return None


def rule_r5(tx: Transaction, base: UserBaseline) -> Optional[RuleHit]:
    """R5 고액 현금 인출: 출금 ≥ 5×avg_withdrawal AND ≥ 100만원"""
    if tx.type != "withdrawal":
        return None
    c = CFG["r5"]
    if tx.amount >= c["multiplier"] * base.avg_withdrawal and tx.amount >= c["absolute_min"]:
        return RuleHit("R5", "고액 현금 인출", c["weight"], "평소보다 큰 현금 인출 (전달책 패턴)")
    return None


def rule_r6(tx: Transaction) -> Optional[RuleHit]:
    """R6 고위험 상품 가입: product_risk_grade 기준"""
    if tx.type != "product":
        return None
    g = tx.product_risk_grade
    if g not in ("mid", "high", "very_high"):
        return None
    reason_map = {
        "mid": "원금 손실 가능성이 있는 상품",
        "high": "원금 손실 위험이 큰 상품",
        "very_high": "구조가 복잡한 초고위험 상품 (ELS·레버리지 등)",
    }
    return RuleHit("R6", "고위험 상품 가입", CFG["r6"][g], reason_map[g])


def rule_r8(tx: Transaction, base: UserBaseline, recent: list[Transaction]) -> Optional[RuleHit]:
    """R8 소비 카테고리 이상: 신규 업종에서 당일 소비 ≥ 5×daily_spend_avg"""
    if tx.type != "payment" or not tx.merchant_category:
        return None
    if tx.merchant_category in base.typical_categories:
        return None
    # 당일 같은 신규 업종 결제 합산 (+ 현재 거래)
    day_spend = sum(
        t.amount for t in recent
        if t.type == "payment" and t.merchant_category == tx.merchant_category
    ) + tx.amount
    if day_spend >= CFG["r8"]["multiplier"] * base.daily_spend_avg:
        return RuleHit("R8", "소비 카테고리 이상", CFG["r8"]["weight"], "평소와 다른 소비 패턴")
    return None


def evaluate_rules(tx: Transaction, base: UserBaseline, recent: list[Transaction]) -> list[RuleHit]:
    """개별 규칙 전체 평가"""
    hits = [
        rule_r1(tx, base),
        rule_r2(tx, base),
        rule_r3(tx, base),
        rule_r4(tx, recent),
        rule_r5(tx, base),
        rule_r6(tx),
        rule_r8(tx, base, recent),
    ]
    return [h for h in hits if h is not None]


# ---------- combo rules ----------


def _combo_c1(hits: list[RuleHit]) -> Optional[ComboHit]:
    """C1: R1(r≥10) & R2 → +25, 최소 High"""
    r1 = _find(hits, "R1")
    r2 = _find(hits, "R2")
    r = r1.meta.get("r", 0) if r1 else 0
    if r1 and r2 and r >= CFG["combo"]["c1"]["r_threshold"]:
        return ComboHit("C1", CFG["combo"]["c1"]["bonus"], "고액 이체 + 신규 수취인 (전형적 사기 패턴)", force_grade="High")
    return None


def _combo_c2(hits: list[RuleHit], base: UserBaseline, recent: list[Transaction]) -> Optional[ComboHit]:
    """
    C2: R2(신규계좌 이체) & R5(고액 현금인출) → 즉시 High + 보호자 즉시 알림.
    두 신호는 보통 서로 다른 거래(이체 후 인출). 현재 거래가 고액 인출(R5)이고
    오늘 앞선 이력에 신규계좌 이체가 있으면 발동.
    """
    if not _find(hits, "R5"):
        return None
    recent_new_payee = any(
        t.type == "transfer" and t.payee_account and t.payee_account not in base.known_payees
        for t in recent
    )
    if recent_new_payee:
        return ComboHit("C2", CFG["combo"]["c2"]["bonus"], "신규 계좌 이체 후 고액 현금인출 (인출·전달책 패턴)",
                        force_grade="High", guardian_alert=True, hold_recommended=True)
    return None


def _combo_c3(hits: list[RuleHit]) -> Optional[ComboHit]:
    """C3: R3 & R1 & R2 → 즉시 High (심야 원격 조작 정황)"""
    if _find(hits, "R3") and _find(hits, "R1") and _find(hits, "R2"):
        return ComboHit("C3", CFG["combo"]["c3"]["bonus"], "심야 시간대 신규 계좌 고액 이체 (원격 조작 정황)",
                        force_grade="High", hold_recommended=True)
    return None


def evaluate_combos(hits: list[RuleHit], tx: Transaction, base: UserBaseline, recent: list[Transaction]) -> list[ComboHit]:
    """조합 규칙 전체 평가"""
    combos = [_combo_c1(hits), _combo_c2(hits, base, recent), _combo_c3(hits)]
    return [c for c in combos if c is not None]


# ---------- utils ----------


def _find(hits: list[RuleHit], rule_id: str) -> Optional[RuleHit]:
    return next((h for h in hits if h.id == rule_id), None)


def _local_hour(iso: str) -> int:
    """ISO 문자열에 적힌 그대로의 '시(hour)' 추출 → 런타임 TZ 영향 회피"""
    m = re.search(r"T(\d{2}):", iso)
    return int(m.group(1)) if m else datetime.fromisoformat(iso).hour


def _epoch(iso: str) -> float:
    return datetime.fromisoformat(iso).timestamp()

File: model/rules.ts
// ============================================================
// 탐지 규칙 구현 (규칙 문서 [2] 개별 규칙, [3] 조합 규칙)
// 각 규칙 = (조건 + 임계값) → RuleHit | null
// 임계값/가중치는 config.ts에서 주입.
// ============================================================

import { RISK_CONFIG as CFG } from "./config";
import { Transaction, UserBaseline, RuleHit, ComboHit } from "./types";

// ---------- 개별 규칙 ----------

/** R1 고액 이체: r = amount / avgTransfer */
export function ruleR1(tx: Transaction, base: UserBaseline): RuleHit | null {
  if (tx.type !== "transfer") return null;
  if (base.avgTransfer <= 0) return null;
  const r = tx.amount / base.avgTransfer;
  const c = CFG.r1;
  let weight = 0;
  let reason = "";
  if (r >= c.t3) { weight = c.w3; reason = "평소 대비 극단적 고액 이체 (사기 전형)"; }
  else if (r >= c.t2) { weight = c.w2; reason = "평소 이체액의 7배 이상"; }
  else if (r >= c.t1) { weight = c.w1; reason = "평소보다 큰 금액 이체"; }
  else return null;
  return { id: "R1", name: "고액 이체", weight, reason, meta: { r } };
}

/** R2 신규 수취인: 이체 수취계좌가 knownPayees에 없음 */
export function ruleR2(tx: Transaction, base: UserBaseline): RuleHit | null {
  if (tx.type !== "transfer") return null;
  if (!tx.payeeAccount) return null;
  if (base.knownPayees.includes(tx.payeeAccount)) return null;
  return { id: "R2", name: "신규 수취인", weight: CFG.r2.weight, reason: "처음 거래하는 계좌" };
}

/** R3 비활동시간 거래: 거래 시각이 activeHours 밖 */
export function ruleR3(tx: Transaction, base: UserBaseline): RuleHit | null {
  const hour = localHour(tx.timestamp);
  const [start, end] = base.activeHours;
  const active =
    start <= end ? hour >= start && hour <= end : hour >= start || hour <= end;
  if (active) return null;
  return { id: "R3", name: "비활동시간 거래", weight: CFG.r3.weight, reason: "평소 거래하지 않는 시간대" };
}

/** R4 단기 연속 거래: 이체/출금 반복. 최댓값 1건만 적용 */
export function ruleR4(tx: Transaction, recent: Transaction[]): RuleHit | null {
  const isSend = (t: Transaction) => t.type === "transfer" || t.type === "withdrawal";
  if (!isSend(tx)) return null;

  const now = new Date(tx.timestamp).getTime();
  const priorMs = recent.filter(isSend).map((t) => new Date(t.timestamp).getTime());
  // 현재 거래 포함 카운트 (recent는 현재 거래 미포함이므로 +1)
  const countWithin = (min: number) =>
    priorMs.filter((t) => now - t >= 0 && now - t <= min * 60_000).length + 1;

  const c = CFG.r4;
  if (countWithin(c.long.windowMin) >= c.long.count)
    return { id: "R4", name: "단기 연속 거래", weight: c.long.weight, reason: "30분 내 5건 이상 반복 송·출금" };
  if (countWithin(c.short.windowMin) >= c.short.count)
    return { id: "R4", name: "단기 연속 거래", weight: c.short.weight, reason: "10분 내 3건 이상 반복 송·출금" };
  return null;
}

/** R5 고액 현금 인출: 출금 ≥ 5×avgWithdrawal AND ≥ 100만원 */
export function ruleR5(tx: Transaction, base: UserBaseline): RuleHit | null {
  if (tx.type !== "withdrawal") return null;
  const c = CFG.r5;
  if (tx.amount >= c.multiplier * base.avgWithdrawal && tx.amount >= c.absoluteMin)
    return { id: "R5", name: "고액 현금 인출", weight: c.weight, reason: "평소보다 큰 현금 인출 (전달책 패턴)" };
  return null;
}

/** R6 고위험 상품 가입: productRiskGrade 기준 */
export function ruleR6(tx: Transaction): RuleHit | null {
  if (tx.type !== "product") return null;
  const g = tx.productRiskGrade;
  if (!g || g === "low" || g === "none") return null;
  const weight = CFG.r6[g];
  if (weight == null) return null;
  const reasonMap: Record<string, string> = {
    mid: "원금 손실 가능성이 있는 상품",
    high: "원금 손실 위험이 큰 상품",
    very_high: "구조가 복잡한 초고위험 상품 (ELS·레버리지 등)",
  };
  return { id: "R6", name: "고위험 상품 가입", weight, reason: reasonMap[g] };
}

/** R7 소비 카테고리 이상: 신규 업종에서 당일 소비 ≥ 5×dailySpendAvg */
export function ruleR7(tx: Transaction, base: UserBaseline, recent: Transaction[]): RuleHit | null {
  if (tx.type !== "payment") return null;
  if (!tx.merchantCategory) return null;
  if (base.typicalCategories.includes(tx.merchantCategory)) return null;

  // 당일 같은 신규 업종 결제 합산 (+ 현재 거래)
  const daySpend =
    recent
      .filter((t) => t.type === "payment" && t.merchantCategory === tx.merchantCategory)
      .reduce((s, t) => s + t.amount, 0) + tx.amount;

  if (daySpend >= CFG.r7.multiplier * base.dailySpendAvg)
    return { id: "R7", name: "소비 카테고리 이상", weight: CFG.r7.weight, reason: "평소와 다른 소비 패턴" };
  return null;
}

/** 개별 규칙 전체 평가 */
export function evaluateRules(
  tx: Transaction,
  base: UserBaseline,
  recent: Transaction[]
): RuleHit[] {
  return [
    ruleR1(tx, base),
    ruleR2(tx, base),
    ruleR3(tx, base),
    ruleR4(tx, recent),
    ruleR5(tx, base),
    ruleR6(tx),
    ruleR7(tx, base, recent),
  ].filter((h): h is RuleHit => h !== null);
}

// ---------- 조합 규칙 ----------

/** C1: R1(r≥10) & R2 → +25, 최소 High */
function comboC1(hits: RuleHit[]): ComboHit | null {
  const r1 = hits.find((h) => h.id === "R1");
  const r2 = hits.find((h) => h.id === "R2");
  const r = (r1?.meta?.r as number) ?? 0;
  if (r1 && r2 && r >= CFG.combo.c1.rThreshold)
    return { id: "C1", bonus: CFG.combo.c1.bonus, forceGrade: "High", reason: "고액 이체 + 신규 수취인 (전형적 사기 패턴)" };
  return null;
}

/**
 * C2: R2(신규계좌 이체) & R5(고액 현금인출) → 즉시 High + 보호자 즉시 알림.
 * 두 신호는 보통 서로 다른 거래에 걸림(이체 후 인출). 현재 거래가 고액 인출(R5)이고,
 * 오늘 앞선 이력에 신규계좌 이체가 있으면 발동.
 */
function comboC2(hits: RuleHit[], tx: Transaction, base: UserBaseline, recent: Transaction[]): ComboHit | null {
  const r5 = hits.find((h) => h.id === "R5");
  if (!r5) return null;
  const recentNewPayeeTransfer = recent.some(
    (t) => t.type === "transfer" && !!t.payeeAccount && !base.knownPayees.includes(t.payeeAccount)
  );
  if (recentNewPayeeTransfer)
    return { id: "C2", bonus: CFG.combo.c2.bonus, forceGrade: "High", guardianAlert: true, holdRecommended: true, reason: "신규 계좌 이체 후 고액 현금인출 (인출·전달책 패턴)" };
  return null;
}

/** C3: R3 & R1 & R2 → 즉시 High (심야 원격 조작 정황) */
function comboC3(hits: RuleHit[]): ComboHit | null {
  const has = (id: string) => hits.some((h) => h.id === id);
  if (has("R3") && has("R1") && has("R2"))
    return { id: "C3", bonus: CFG.combo.c3.bonus, forceGrade: "High", holdRecommended: true, reason: "심야 시간대 신규 계좌 고액 이체 (원격 조작 정황)" };
  return null;
}

/** 조합 규칙 전체 평가 */
export function evaluateCombos(
  hits: RuleHit[],
  tx: Transaction,
  base: UserBaseline,
  recent: Transaction[]
): ComboHit[] {
  return [comboC1(hits), comboC2(hits, tx, base, recent), comboC3(hits)].filter(
    (c): c is ComboHit => c !== null
  );
}

// ---------- 유틸 ----------

/** ISO 문자열에 적힌 그대로의 '시(hour)' 추출 → 런타임 TZ 영향 회피 */
function localHour(iso: string): number {
  const m = iso.match(/T(\d{2}):/);
  return m ? parseInt(m[1], 10) : new Date(iso).getHours();
}
